1. 회귀분석


1-1. 데이터를 8:2로 분할하고 선형회귀를 적용하시오. 결정계수와 rmse를 구하시오

In [1]:
import pandas as pd
import numpy as np
import mglearn

x,y = mglearn.datasets.load_extended_boston()

c:\Users\suhye\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (

In [2]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2,random_state=0)

In [3]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(x_train, y_train)
lr_pred = lr.predict(x_test)

In [4]:
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_test,lr_pred)
rmse = np.sqrt(mse)
lr_score = lr.score(x_test,y_test)

print("결정계수 : ",lr_score)
print('RMSE ', rmse)

결정계수 :  0.6158858584078815
RMSE  5.592657237078594


1-2. 데이터를 8:2로 분할하고 릿지 회귀를 적용하시오. alpha값을 0부터 1까지 0.1단위로 모두 탐색해서 결정계수가 가장 높을 때의 알파를 찾고, 해당 알파로 다시 모델을 학습해서 결정계수와 rmse를 계산하시오. 

In [5]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error

ridge = Ridge(normalize=True)
alpha = np.arange(0,1,0.1)
param_grid = {'alpha':alpha}

ridge_model = GridSearchCV(ridge,param_grid)
ridge_model.fit(x_train,y_train)

print(ridge_model.best_params_)
print('릿지 회귀 결정계수 :',ridge_model.best_estimator_.score(x_test,y_test))
print('릿지 회귀 RMSE : ',np.sqrt(mean_squared_error(y_test,ridge_model.best_estimator_.predict(x_test))))

c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Ridge())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * n_samples. 
  warnings.warn(
c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklea

{'alpha': 0.1}
릿지 회귀 결정계수 : 0.7356090234954484
릿지 회귀 RMSE :  4.639929365714194


c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Ridge())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * n_samples. 
  warnings.warn(
c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklea

릿지 회귀에서는 , GridSearchCV 함수를 활용하여 alpha를 0부터 1까지 적용시켜 최적의 모델을 찾아냈다. 최적의 모델을 갖는 alpha값은 0.1이었다. alpha값을 조정하여 모델의 학습 정도를 조절할 수 있으며, 평가지표로는 R2와 RMSE가 사용된다. alpha=1의 모델로 test 데이터를 예측하여 검증해본 결과, R2는 약 0.735, RMSE는 약 4.6으로 나타났다. 

1-3. 데이터를 8:2로 분할하고 라쏘 회귀를 적용하시오. alpha값을 0부터 1까지 0.1단위로 모두 탐색해서 결정계수가 가장 높을 대의 알파를 찾고, 해당 알파로 다시 모델을 학습해서 결정계수와 rmse를 계산하시오.

In [6]:
from sklearn.linear_model import Lasso

lasso = Lasso(normalize=True)
alpha = np.arange(0,1,0.1)
param_grid = {'alpha':alpha}
lasso_model = GridSearchCV(lasso,param_grid)
lasso_model.fit(x_train,y_train)

lasso_score = lasso_model.best_estimator_.score(x_test,y_test)
lasso_pred = lasso_model.best_estimator_.predict(x_test)
rmse = np.sqrt(mean_squared_error(y_test,lasso_pred))

print(lasso_model.best_params_)
print('라쏘 회귀 결정계수: ',lasso_score)
print('RMSE :',rmse)

c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Lasso())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * np.sqrt(n_samples). 
  warnings.warn(
c:\Users\suhye\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:680: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  estimator.fit(X_train, y_train, **fit_params)
c:\Users\suhye\anaconda3\lib\site-pack

{'alpha': 0.0}
라쏘 회귀 결정계수:  0.6901880385280229
RMSE : 5.022698918446971


c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Lasso())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * np.sqrt(n_samples). 
  warnings.warn(
c:\Users\suhye\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning: 'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

f

라쏘회귀에서도 GridSearchCV 함수를 사용하여 alpha를 0부터 1까지 적용시켜 최적의 모델을 찾아냈다. 최적의 모델을 갖는 알파값은 0이었다. alpha=0의 모델로 test 데이터를 예측하여 검증해본 결과, R2는 약 0.69, RMSE는 약 5.0으로 나타났다. 해당 데이터에서는 Ridge 회귀의 예측력이 더 좋은 것을 알 수 있었다. 